src/nfa/intersection.cc

```diff
@@ -16,177 +16,233 @@
+ */
+    auto create_product_state_and_symbol_post = [&](const State lhs_target, const State rhs_target, SymbolPost& product_symbol_post)
+    {
+        State product_target = get_state_from_product_storage(lhs_target, rhs_target );
 
-namespace mata::nfa {
+        if ( product_target == Limits::max_state )
+        {
+            product_target = product.add_state();
+            assert(product_target < Limits::max_state);
 
-Nfa intersection(const Nfa& lhs, const Nfa& rhs, bool preserve_epsilon,
-                 std::unordered_map<std::pair<State,State>, State> *prod_map) {
+            insert_to_product_storage(lhs_target,rhs_target, product_target);
 
-    const std::set<Symbol> epsilons({EPSILON});
-    return algorithms::intersection_eps(lhs, rhs, preserve_epsilon, epsilons, prod_map);
-}
+            worklist.push_back(product_target);
 
-Nfa mata::nfa::algorithms::intersection_eps(
-        const Nfa& lhs, const Nfa& rhs, bool preserve_epsilon, const std::set<Symbol>& epsilons,
-        std::unordered_map<std::pair<State, State>, State> *prod_map) {
-    Nfa product{}; // Product of the intersection.
-    // Product map for the generated intersection mapping original state pairs to new product states.
-    std::unordered_map<std::pair<State,State>, State> product_map{};
-    std::pair<State,State> pair_to_process{}; // State pair of original states currently being processed.
-    std::deque<std::pair<State,State>> pairs_to_process{}; // Set of state pairs of original states to process.
+            if (final_condition(lhs_target,rhs_target)) {
+                product.final.insert(product_target);
+            }
+        }
+        //TODO: Push_back all of them and sort at the could be faster.
+        product_symbol_post.insert(product_target);
+    };
 
     // Initialize pairs to process with initial state pairs.
     for (const State lhs_initial_state : lhs.initial) {
         for (const State rhs_initial_state : rhs.initial) {
             // Update product with initial state pairs.
-            const std::pair<State,State> this_and_other_initial_state_pair(lhs_initial_state, rhs_initial_state);
-            const State new_intersection_state = product.add_state();
-
-            product_map[this_and_other_initial_state_pair] = new_intersection_state;
-            pairs_to_process.push_back(this_and_other_initial_state_pair);
-
-            product.initial.insert(new_intersection_state);
-            if (lhs.final[lhs_initial_state] && rhs.final[rhs_initial_state]) {
-                product.final.insert(new_intersection_state);
+            const State product_initial_state = product.add_state();
+            insert_to_product_storage(lhs_initial_state,rhs_initial_state,product_initial_state);
+            worklist.push_back(product_initial_state);
+            product.initial.insert(product_initial_state);
+            if (final_condition(lhs_initial_state,rhs_initial_state)) {
```

Comment:
I would name the lamba `is_final` or `are_final` or maybe `final_test`? It's shorter. Final condition does not sound much like function.

Reply:
I still like the final_condition, is that ok?

Reply:
Yeah, I don't really give a fuck. I'm simply suggesting alternatives.

Reply:
ok